tpu-client-next/src/node_address_service/leader_tpu_cache_service.rs

```diff
@@ -0,0 +1,520 @@
+            *num_failures = num_failures.saturating_add(1);
+            warn!("Failed to update {label}: {e} ({num_failures} consecutive failures)",);
+
+            if *num_failures >= max_failures {
+                error!("Max consecutive failures for {label}, giving up.");
+                Err(e)
+            } else {
+                Ok(())
+            }
+        }
+    }
+}
+
+#[derive(PartialEq, Debug)]
+struct LeaderTpuMap {
+    last_cluster_refresh: Instant,
+    leader_tpu_map: HashMap<Pubkey, SocketAddr>,
+}
+
+impl LeaderTpuMap {
+    async fn new(rpc_client: &impl ClusterInfoProvider) -> Result<Self, Error> {
+        let leader_tpu_map = rpc_client.leader_tpu_map().await?;
+        Ok(Self {
+            last_cluster_refresh: Instant::now(),
+            leader_tpu_map,
+        })
+    }
+
+    fn get(&self, leader: &Pubkey) -> Option<&SocketAddr> {
+        self.leader_tpu_map.get(leader)
+    }
+}
+
+#[derive(PartialEq, Debug)]
+struct SlotLeaders {
+    first_slot: Slot,
+    leaders: Vec<Pubkey>,
+}
+
+impl SlotLeaders {
+    async fn new(
+        rpc_client: &impl ClusterInfoProvider,
+        estimated_current_slot: Slot,
+        slots_in_epoch: Slot,
+    ) -> Result<Self, Error> {
+        Ok(Self {
+            first_slot: estimated_current_slot,
+            leaders: rpc_client
+                .slot_leaders(estimated_current_slot, slots_in_epoch)
+                .await?,
+        })
+    }
+
+    fn last_slot(&self) -> Slot {
+        self.first_slot + self.leaders.len().saturating_sub(1) as u64
+    }
+
+    fn slot_leader(&self, slot: Slot) -> Option<&Pubkey> {
+        slot.checked_sub(self.first_slot)
+            .and_then(|index| self.leaders.get(index as usize))
+    }
+
+    fn is_last_slot_in_window(&self, slot: Slot) -> Option<bool> {
+        slot.checked_sub(self.first_slot).and_then(|index| {
+            let index = index as usize;
+            if index + 1 < self.leaders.len() {
+                Some(self.leaders[index] != self.leaders[index + 1])
+            } else {
+                None
+            }
+        })
+    }
+}
+
+#[derive(PartialEq, Debug)]
+struct EpochInfo {
+    slots_in_epoch: Slot,
+    last_slot_in_epoch: Slot,
+}
+
+impl EpochInfo {
+    async fn new(
+        rpc_client: &impl ClusterInfoProvider,
+        estimated_current_slot: Slot,
+    ) -> Result<Self, Error> {
+        let (slots_in_epoch, last_slot_in_epoch) =
+            rpc_client.epoch_info(estimated_current_slot).await?;
+        Ok(Self {
+            slots_in_epoch,
+            last_slot_in_epoch,
+        })
+    }
+}
+
+#[async_trait]
+impl ClusterInfoProvider for RpcClient {
+    async fn leader_tpu_map(&self) -> Result<HashMap<Pubkey, SocketAddr>, Error> {
+        let cluster_nodes = self.get_cluster_nodes().await.map_err(Error::RpcError)?;
+        Ok(extract_cluster_tpu_sockets(cluster_nodes))
+    }
+
+    async fn epoch_info(&self, estimated_current_slot: Slot) -> Result<(Slot, Slot), Error> {
```

Comment:
nit: I would not leak that `estimated_current_slot` into generic parts of the systems, i.e. those parts which are not interested in weather the slot is the current slot or not

Reply:
agree.